interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,384 @@
+func (pdb *PromptDB) nextID() string {
+	pdb.maxID++
+	idStr := fmt.Sprintf("%016X", pdb.maxID)
+	osutil.AtomicWriteFile(pdb.maxIDPath, []byte(idStr), 0600, 0)
+	return idStr
+}
+
+// AddOrMerge checks if the given prompt contents are identical to an existing
+// prompt and, if so, merges with it by adding the given listenerReq to it.
+// Otherwise, adds a new prompt with the given contents to the prompt DB.
+//
+// If the prompt was merged with an identical existing prompt, returns the
+// existing prompt and true, indicating it was merged. If a new prompt was
+// added, returns the new prompt and false, indicating the prompt was not
+// merged.
+//
+// The caller must ensure that the given permissions are in the order in which
+// they appear in the available permissions list for the given interface.
+func (pdb *PromptDB) AddOrMerge(metadata *prompting.Metadata, path string, permissions []string, listenerReq *listener.Request) (*Prompt, bool) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[metadata.User]
+	if !ok {
+		pdb.perUser[metadata.User] = &userPromptDB{
+			ByID: make(map[string]*Prompt),
+		}
+		userEntry = pdb.perUser[metadata.User]
+	}
+
+	availablePermissions, _ := prompting.AvailablePermissions(metadata.Interface)
+	// Error should be impossible, since caller has already validated that iface
+	// is valid, and tests check that all valid interfaces have valid available
+	// permissions returned by AvailablePermissions.
+
+	constraints := &PromptConstraints{
+		Path:                 path,
+		Permissions:          permissions,
+		AvailablePermissions: availablePermissions,
+	}
+
+	// Search for an identical existing prompt, merge if found
+	for _, prompt := range userEntry.ByID {
+		if prompt.Snap == metadata.Snap && prompt.Interface == metadata.Interface && prompt.Constraints.equals(constraints) {
+			prompt.listenerReqs = append(prompt.listenerReqs, listenerReq)
+			// Although the prompt itself has not changed, re-record a notice
+			// to re-notify clients to respond to this request. A client may
+			// have replied with a malformed response and not retried after
+			// receiving the error, so this notice encourages it to try again
+			// if the user retries the operation.
+			pdb.notifyPrompt(metadata.User, prompt.ID, nil)
+			return prompt, true
+		}
+	}
+
+	id := pdb.nextID()
+	timestamp := time.Now()
+	prompt := &Prompt{
+		ID:           id,
+		Timestamp:    timestamp,
+		Snap:         metadata.Snap,
+		Interface:    metadata.Interface,
+		Constraints:  constraints,
+		listenerReqs: []*listener.Request{listenerReq},
+	}
+	userEntry.ByID[id] = prompt
+	pdb.notifyPrompt(metadata.User, id, nil)
+	return prompt, false
+}
+
+// Prompts returns a slice of all outstanding prompts.
+func (pdb *PromptDB) Prompts(user uint32) []*Prompt {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return []*Prompt{}
```

Comment:
Can you just return nil here. It is equivalent.

Reply:
Not quite equivalent in this case... the output of this function is directly marshalled and sent over an API response, and `[]*Prompt{}` marshals to `"[]"` whereas `nil` marshals to `null`, and we want the former.

Reply:
Fixed, I'll normalize in the daemon.